frontend/src/core/codemirror/readonly/extension.ts

```diff
@@ -0,0 +1,187 @@
+} from "@/core/websocket/connection-utils";
+
+/**
+ * State effect for updating readonly status based on connection
+ */
+const updateReadonlyEffect = StateEffect.define<boolean>();
+
+/**
+ * Compartment for managing readonly configuration
+ */
+const readonlyCompartment = new Compartment();
+
+// Floating indicator plugin
+const INDICATOR_ID = "marimo-connecting-indicator";
+function getOrCreateIndicator(): HTMLDivElement {
+  let indicator = document.getElementById(
+    INDICATOR_ID,
+  ) as HTMLDivElement | null;
+  if (!indicator) {
+    indicator = document.createElement("div");
+    indicator.id = INDICATOR_ID;
+    indicator.textContent = "Connecting…";
+    indicator.style.position = "fixed";
+    indicator.style.zIndex = "9999";
+    indicator.style.background = "#222";
+    indicator.style.color = "#fff";
+    indicator.style.padding = "2px 8px";
+    indicator.style.borderRadius = "6px";
+    indicator.style.fontSize = "0.9em";
+    indicator.style.boxShadow = "0 2px 8px rgba(0,0,0,0.15)";
+    indicator.style.pointerEvents = "none";
+    indicator.style.display = "none";
+    document.body.append(indicator);
+  }
+  return indicator;
+}
+
+function showIndicatorAt(left: number, top: number) {
+  const indicator = getOrCreateIndicator();
+  indicator.style.left = `${left}px`;
+  indicator.style.top = `${top}px`;
+  indicator.style.display = "block";
+  indicator.textContent = "Connecting…";
+}
+function hideIndicator() {
+  const indicator = document.getElementById(INDICATOR_ID);
+  if (indicator) {
+    indicator.style.display = "none";
+  }
+}
+
+let hideIndicatorTimeout: number | null = null;
+
+/**
+ * Creates a dynamic readonly extension that toggles based on WebSocket connection status.
+ * When disconnected or connecting, the editor becomes readonly.
+ * When connected, the editor becomes editable.
+ */
+export function dynamicReadonly(store: ReturnType<typeof createStore>) {
+  const getConnectionState = () => store.get(connectionAtom).state;
+
+  // Initial readonly state - start as readonly until connected
+  let isReadonly = !isAppConnected(getConnectionState());
+
+  const connectingIndicatorPlugin = ViewPlugin.define((view) => {
+    let listenersAttached = false;
+    const dom = view.dom;
+
+    function showNearCaret() {
+      const sel = view.state.selection.main;
+      const coords = view.coordsAtPos(sel.head);
+      if (coords) {
+        showIndicatorAt(coords.left + 2, coords.bottom + 2);
+      } else {
+        showIndicatorAt(window.innerWidth / 2, window.innerHeight * 0.1);
+      }
+      if (hideIndicatorTimeout) {
+        window.clearTimeout(hideIndicatorTimeout);
+      }
+      hideIndicatorTimeout = window.setTimeout(hideIndicator, 1500);
+    }
+
+    function onUserInput() {
+      showNearCaret();
+    }
+
+    function attachListeners() {
+      if (!listenersAttached) {
+        dom.addEventListener("keydown", onUserInput, true);
+        dom.addEventListener("paste", onUserInput, true);
```

Comment:
I prefer the options object because it's slightly more explicit what the boolean is. 

```suggestion
        dom.addEventListener("keydown", onUserInput, { capture: true });
        dom.addEventListener("paste", onUserInput, { capture: true });
```